### PR TITLE
Post Mortem

### PR DESCRIPTION
**Pull request by ZenGround0**
@Stebalien this is about ready but could use a quick look over before merging.

### CODE REVIEW

**Comment by ZenGround0** [comment]
The <a/> sections rendered invisible on my local editor. Figuring out how to keep them hidden with some other approach for internal links.

File: docs/post-mortem-jan-21.md [hunk 1]
```diff
@@ -0,0 +1,246 @@
...
+- [Part 2: Discussion](#disc-C)
+
+[**Summary discussion and recommendations**](#recc)
+
+- [What *has* been working for us?](#working)
+
+
+
+# Intro
+<a name="intro"></a>
```
Comment (Stebalien) [inline docs/post-mortem-jan-21.md:28]: Should be `id`, not name. Also, you _should_ be able to just link to `[Intro](#intro)` (i.e., `my-title-with-dashes-and-no-caps`).

File: docs/post-mortem-jan-21.md [hunk 2]
```diff
@@ -0,0 +1,246 @@
...
+
+
+### Discovery and mitigation
+
+Community contributor [@zgfzgf](https://github.com/zgfzgf) reported the issue to us on 23 October 2020. 
+
+There were no instances of this error being exploited. We quietly mitigated the issue in the release of Lotus 1.1.3 Nov 13 2020, by changing the VerifyConsensusFault syscall implementation to require the offending blocks’ signers to be the worker key of the receiving actor. An exploitation at this point would have caused a fork between participants running 1.1.3 and earlier versions. The mitigation was considered to be in effect only after Lotus 1.2 was released, which included the network version 7 upgrade (code released Nov 18th 2020 and network upgrade at epoch 265200), and thus forced all participants to be running code containing the workaround.
```
Comment (Stebalien) [inline docs/post-mortem-jan-21.md:60]: I'd say that the mitigation wasn't considered to be _fully_ effective. It was still in effect.

File: docs/post-mortem-jan-21.md [hunk 3]
```diff
@@ -0,0 +1,246 @@
...
+*   Non-determinism, including unstable gas usage, is not tested in actors directly
+    *   It is checked in the VM test vectors, but they target the VM itself, and do not cover a wide range of actor behaviour
+*   We could use a “golden file” approach to check for deterministic execution
+    *   This would also be helpful for detecting unintended behaviour changes between network version upgrades.
+*   Mutation testing could have detected this in theory, since the “sorting” line has no effect.
+    *   We’d need to customize a mutation testing library to work well with actors.
+*   The deadline_state_test.go does exercise termination of multiple sectors. This might have caught the issue in combination with a golden file.
+*   Simple fuzz testing isn’t helpful; the bug required building up enough state to terminate multiple sectors. 
+*   Randomisation over high level agent behaviour could have exercised this (in combination with a golden/expectation file).
+*   We’re not aware of anything other than map iteration that could be non-deterministic, but something could be introduced later.
```
Comment (Stebalien) [inline docs/post-mortem-jan-21.md:157]: Can we also mention that we explicitly lint _against_ map iteration?

File: docs/post-mortem-jan-21.md [hunk 4]
```diff
@@ -0,0 +1,246 @@
...
+
+# Multisig Infinite Re-entrancy
+<a name="C"></a>
+
+## Part 1: Facts
+<a name="facts-C"></a>
+
+### Description of the bug
+
+A 1 of n multisig actor can add itself as a signer, propose a message to the multisig actor address such that the proposal message is an approval message to the multisig actor address approving the tx id of the approval message itself. A multisig transaction of this form recurses indefinitely approving the approval message at each iteration.  This is possible because the multisig actor does not remove the tx id from multisig state before launching the send call that executes the transaction.
```
Comment (Stebalien) [inline docs/post-mortem-jan-21.md:170]: past tense?

**Review by Stebalien — APPROVED** [review]

File: docs/post-mortem-jan-21.md (hunk 1 above)
Comment (ZenGround0) [reply docs/post-mortem-jan-21.md:28]: I tried the default heading names. But some of the subsections linked in table of contents had the same name which led me down this path.

Comment (Stebalien) [reply docs/post-mortem-jan-21.md:28]: Ah... it looks like you can handle it by adding `-1`, `-2`, etc. But that's awkward.

**Comment by anorth** [comment]
😍

**Pull request merged**